File: src/python/grid_walk.py
import sys
from functools import wraps
from time import time

def timed(f):
    @wraps(f)
    def wrapper(*args, **kwds):
        start = time()
        result = f(*args, **kwds)
        elapsed = time() - start
        print "%s took %f seconds to finish" % (f.__name__, elapsed)
        return result
    return wrapper

explored = set()
_cache = {}

def sum_of_digits(number):
    n_abs = abs(number)
    n = n_abs
    if n in _cache:
        return _cache[n]
    r = 0
    while n:
        r += n % 10
        n /= 10
    _cache[n_abs] = r
    return r


def can_walk((x, y)):
    return sum_of_digits(x) + sum_of_digits(y) <= 19


def walkable_neighbours((x, y)):
    neighbours = []
    for dx in (-1, 1):
        p = (x + dx, y)
        if (not p in explored) and can_walk(p):
            neighbours.append(p)
    for dy in (-1, 1):
        p = (x, y + dy)
        if (not p in explored) and can_walk(p):
            neighbours.append(p)
    return neighbours

@timed
def grid_walk():
    to_explore = [(0, 0)]
    while to_explore:
        point = to_explore.pop()
        explored.add(point)
        for neighbour in walkable_neighbours(point):
            to_explore.append(neighbour)
    return len(explored)


if __name__ == '__main__':
    print(grid_walk())

File: src/java/grid_walk.java
import java.io.IOException;
import java.util.*;

public class grid_walk {

    private static boolean canWalkOn(final int x, final int y) {
        final int sum = sumOfDigits(x) + sumOfDigits(y);
        return sum <= 19;
    }

    private static int sumOfDigits(int n) {
        int sum = 0;
        n = Math.abs(n);
        while (n > 0) {
            sum = sum + (n % 10);
            n = n / 10;
        }
        return sum;
    }

    private static final int[] OFFSETS = {-1, 1};

    private static List<Point> walkableNeighbours(final Point p) {
        final List<Point> points = new ArrayList<>(4);
        for (final int dx : OFFSETS) {
            final int x = p.x + dx;
            if (canWalkOn(x, p.y)) {
                points.add(new Point(x, p.y));
            }
        }
        for (final int dy : OFFSETS) {
            final int y = p.y + dy;
            if (canWalkOn(p.x, y)) {
                points.add(new Point(p.x, y));
            }
        }
        return points;
    }

    private static int gridWalk() {
        final Set<Point> explored = new HashSet<>();
        final Stack<Point> toExplore = new Stack<>();
        toExplore.push(new Point(0, 0));
        while (!toExplore.isEmpty()) {
            final Point point = toExplore.pop();
            explored.add(point);
            for (final Point neighbour : walkableNeighbours(point)) {
                if (!explored.contains(neighbour)) {
                    toExplore.push(neighbour);
                }
            }
        }
        return explored.size();
    }

    public static void main(String[] args) throws IOException {
        System.out.println(gridWalk());
    }

    static class Point {
        final int x;
        final int y;
        
        final static Map<Integer, Point> cache = new HashMap<>();

        Point(final int x, final int y) {
            this.x = x;
            this.y = y;
        }
        
        public static Point at(final int x, final int y) {
           int key = Objects.hash(x, y);
           Point point = cache.get(key);
           if (point == null) {
               point = new Point(x, y);
               cache.put(key, point);
           }
           return point;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Point point = (Point) o;

            if (x != point.x) return false;
            if (y != point.y) return false;

            return true;
        }

        @Override
        public int hashCode() {
            return Objects.hash(x, y);
        }        
    }

}

